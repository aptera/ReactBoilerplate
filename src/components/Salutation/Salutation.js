import React, {Component} from 'react';
import './salutation.scss';
import { setInterval } from 'timers';

class Salutation extends Component {

    constructor(props){
        super(props);
        
        this.intervalId = 0;
        this.wordsArray = props.words === undefined ? ['Salutations!'] : this.makeWordArray(props.words);
        this._isMounted = false;  // using a flag because: https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html
        let word = this.getRandomWord();
        this.state = {
            word: word
        };
    }

    componentDidMount() {
        this._isMounted = true;
        if (this.props.refreshInterval) {
            this.intervalId = setInterval(this.changeWord.bind(this), this.props.refreshInterval);
        }
    }

    changeWord() {
        if(!this._isMounted)
            return;
        let word = this.getRandomWord();
        this.setState({ word: word });
    }

    makeWordArray(wordsString) {
        return wordsString.split(',').map(function(item) {
            return item.trim();
        });
    }

    componentWillUnmount() {
        this._isMounted = false;
        clearInterval(this.intervalId);
    }

    getRandomWord(){
        return this.wordsArray[Math.floor(Math.random() * this.wordsArray.length)];
    }

    render() {
        return (
            <div className={this.props.class}><a href={'https://www.merriam-webster.com/dictionary/' + this.state.word}>{this.state.word}</a></div>
        );
    }
}

export default Salutation;

