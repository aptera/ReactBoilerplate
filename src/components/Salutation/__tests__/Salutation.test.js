import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import Salutation from '../Salutation';

describe('the salutation component', () => {
    it('renders correctly', () => {
        const tree = renderer.create(
            <Salutation />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('given no greetings', () => {
        it('displays the default greeting', () => {
            let SalutationComponent = ReactTestUtils.renderIntoDocument(<Salutation />);
            let h1 = ReactTestUtils.findRenderedDOMComponentWithTag(SalutationComponent, 'h1');
            expect(h1.textContent).toBe('hello world');
        });
    });

    describe('given a word', () => {
        it('displays that word instead of the default greeting', () => {
            let SalutationComponent = ReactTestUtils.renderIntoDocument(<Salutation words="heyo"/>);
            let h1 = ReactTestUtils.findRenderedDOMComponentWithTag(SalutationComponent, 'h1');
            expect(h1.textContent).toBe('heyo');
        });
    });

    describe('given a comma-seperated list of greetings', () => {
        it('displays one greeting from list', () => {
            let SalutationComponent = ReactTestUtils.renderIntoDocument(<Salutation words="heyo, heyo"/>);
            let h1 = ReactTestUtils.findRenderedDOMComponentWithTag(SalutationComponent, 'h1');
            expect(h1.textContent).toBe('heyo');
        });
    });

});