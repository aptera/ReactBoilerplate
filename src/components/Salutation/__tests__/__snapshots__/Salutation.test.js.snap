// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the salutation component renders correctly 1`] = `
<h1>
  Salutations!
</h1>
`;
